File: python/StaticUtil.py
# Static utility functions. All logs are base 2!

from collections import defaultdict
from math import log

############################################################################

# Given two numbers log(x) and log(y), returns log(x+y). If both log(x) and log(y) are negative,
# alculates the log of the sum in the following manner to reduce the risk of underflow:
# Finds log(z) = min(log(x), log(y)). Then adds log(z) to both log(x) and log(y) before
# exponentiating, adds the two exponentiations, takes the log of the sum, and then
# subtracts log(z). This calculation works properly because z*x + z*y = z*(x+y),
# so 2^(log(z)+log(x)) + 2^(log(z)+log(y)) = 2^(log(z)+log(x+y)).
# Thus, log(2^(log(z)+log(x)) + 2^(log(z)+log(y))) = log(2^(log(z)+log(x+y))) = log(z)+log(x+y).
#
# Note: If logx (logy) is passed as None, then x (y) is interpreted as 0. If both parameters
# are passed as None, then None is returned.
def calc_log_sum_of_logs(logx, logy):
    if logx is None and logy is None:
        return None
    elif logx is None:
        return logy
    elif logy is None:
        return logx

    if logx >= 0 or logy >= 0:
        x = 2 ** logx
        y = 2 ** logy
        sum_of_x_and_y = x + y
        log_sum_of_x_and_y = log(sum_of_x_and_y, 2)
        return log_sum_of_x_and_y

    logz = min(logx, logy)
    scaled_logx = logx + logz
    scaled_logy = logy + logz

    scaled_x = 2 ** scaled_logx
    scaled_y = 2 ** scaled_logy

    scaled_sum_of_x_and_y = scaled_x + scaled_y
    log_scaled_sum_of_x_and_y = log(scaled_sum_of_x_and_y, 2)
    log_sum_of_x_and_y = log_scaled_sum_of_x_and_y - logz

    return log_sum_of_x_and_y

############################################################################

# Returns logx + logy = log(x*y).

# Note: If logx (logy) is passed as None, then x (y) is interpreted as 0. If either x or y is 0,
# then None is returned to indicate that x*y = 0.
def calc_log_prod_of_logs(logx, logy):
    if logx is None or logy is None:
        return None

    return logx + logy

############################################################################

# Parses the given transition probabilities file for the hidden markov model.
# Each line of the file must be of the form "<from_state> <to_state> <prob>". The from_state
# of the first line must be a token representing the start of an emission sequence. Returns
# (a dictionary mapping each (from_state, to_state) tuple to the probability of the transition,
# the set of all states, the emission sequence start token).
# Transitions not present in the file will be given 0 probability.
def parse_trans_prob_file(trans_prob_file):
    trans_prob_dict = defaultdict(int)
    state_set = set()
    start_token = None

    for i, unstripped_line in enumerate(trans_prob_file):
        line = unstripped_line.rstrip()
        if line:
            tokens = line.split()

            if len(tokens) != 3:
                raise BaseException("Line %d of the transition probabilities file has an invalid format.\n" % (i+1))

            from_state = tokens[0]
            to_state = tokens[1]

            if i == 0:
                start_token = from_state
            else:
                if to_state is start_token:
                    raise BaseException("Line %d of the transition probabilities file has a transition to the start token, which is invalid.\n" % (i+1))

            try:
                prob = float(tokens[2])
            except ValueError:
                raise BaseException("On line %d of the transition probabilities file, the probability %s is not a valid float.\n" % (i+1, tokens[2]))

            trans_prob_dict[(from_state, to_state)] = prob
            state_set.add(from_state)
            state_set.add(to_state)

    if start_token is None:
        raise BaseException("The transition probabilities file is empty.\n")

    state_set.remove(start_token)
    return (trans_prob_dict, state_set, start_token)

############################################################################

# Parses the given emission probabilities file for the hidden markov model.
# Each line of the file must be of the form "<state> <emitted_token> <prob>". Returns
# a dictionary mapping each (state, token) tuple to the probability of the emission.
# Emissions not present in the file will be given 0 probability.
def parse_emis_prob_file(emis_prob_file):
    emis_prob_dict = defaultdict(int)
    empty = True

    for i, unstripped_line in enumerate(emis_prob_file):
        line = unstripped_line.rstrip()
        if line:
            empty = False

            tokens = line.split()

            if len(tokens) != 3:
                raise BaseException("Line %d of the emission probabilities file has an invalid format.\n" % (i+1))

            state = tokens[0]
            emitted_token = tokens[1]

            try:
                prob = float(tokens[2])
            except ValueError:
                raise BaseException("On line %d of the emission probabilities file, the probability %s is not a valid float.\n" % (i+1, tokens[2]))

            emis_prob_dict[(state, emitted_token)] = prob

    if empty:
        raise BaseException("The emission probabilities file is empty.\n")

    return emis_prob_dict

############################################################################

# Returns a list of dictionaries, where each dictionary at index i corresponds to the observation
# w_i. Each entry in each dictionary maps a state X to the log sum over all possible previous state
# taggings of the probability of w_i given this state X.
def calculate_forward_matrix(trans_prob_dict, emis_prob_dict, state_set, start_token, obs_list):
    forward_log_prob_dicts = []

    for i, obs in enumerate(obs_list):
        forward_log_prob_dict = defaultdict(lambda: None)
        forward_log_prob_dicts.append(forward_log_prob_dict)

        prev_state = start_token
        for state in state_set:
            # P(state|prev_state) * P(obs|state)
            log_prob_state_given_prevstate = trans_prob_dict[(prev_state, state)]



############################################################################

File: src/main/java/edu/dartmouth/hmmem/StaticUtil.java
package edu.dartmouth.hmmem;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.apache.hadoop.fs.FSDataInputStream;

/**
 * 静态工具类，包含了分布式HMM-EM的相关方法
 */
public class StaticUtil {

	/**
	 * Given two numbers log(x) and log(y), returns log(x+y). If both log(x) and log(y) are negative,
	 * calculates the log of the sum in the following manner to reduce the risk of underflow:
	 * Finds log(z) = min(log(x), log(y)). Then adds log(z) to both log(x) and log(y) before
	 * exponentiating, adds the two exponentiations, takes the log of the sum, and then
	 * subtracts log(z). This calculation works properly because z*x + z*y = z*(x+y),
	 * so 2^(log(z)+log(x)) + 2^(log(z)+log(y)) = 2^(log(z)+log(x+y)).
	 * Thus, log(2^(log(z)+log(x)) + 2^(log(z)+log(y))) = log(2^(log(z)+log(x+y))) = log(z)+log(x+y).
	 *
	 * Note: If logx (logy) is passed as null, then x (y) is interpreted as 0. If both parameters
	 * are passed as null, then null is returned.
	 */
	public static Double calcLogSumOfLogs(Double logX, Double logY) {
		if (logX == null && logY == null) {
			return null;
		} else if (logX == null) {
			return logY;
		} else if (logY == null) {
			return logX;
		}

		if (logX >= 0 || logY >= 0) {
			double x = Math.pow(2, logX);
			double y = Math.pow(2, logY);

			double logSum = Math.log(x + y) / Math.log(2);
			return logSum;
		}

		double logZ = Math.max(logX, logY) * -1.0;
		double scaledLogX = logX + logZ;
		double scaledLogY = logY + logZ;

		double scaledX = Math.pow(2, scaledLogX);
		double scaledY = Math.pow(2, scaledLogY);

		double logScaledSum = Math.log(scaledX + scaledY) / Math.log(2);
		double logSum = logScaledSum - logZ;

		if (!(Double.isNaN(logX) || Double.isNaN(logY))) {
			System.out.println("logX = " + logX + ", logY = " + logY + ", logSum = " + logSum);
		}

		return logSum;
	}

	/**
	 * Given two numbers log(x) and log(y), returns log(x*y) = log(x) + log(y). This method is used
	 * because it handles the case where x or y is 0, which is represented by log(x) or log(y) being null.
	 * In this case, x*y == 0, so null is returned to represent log(0).
	 */
	public static Double calcLogProductOfLogs(Double logX, Double logY) {
		if (logX == null || logY == null) {
			return null;
		}

		return logX + logY;
	}

	/**
	 * Normalizes a log probability map such that all the probabilities
	 * where the first string in the string pair key is some string x sum to 1.0.
	 */
	public static void normalizeLogProbMap(Map<StringPair, Double> logProbMap) {
		Map<String, Double> logProbSumMap = new HashMap<>();

		// See how much we have to scale down by to normalize.
		for (StringPair stringPair : logProbMap.keySet()) {
			String x = stringPair.getX();

			Double logProb = logProbMap.get(stringPair);
			Double prevLogProbSum = logProbSumMap.get(x);

			Double newLogProbSum = calcLogSumOfLogs(logProb, prevLogProbSum);
			logProbSumMap.put(x, newLogProbSum);
		}

		// Do the normalization.
		for (StringPair stringPair : logProbMap.keySet()) {
			Double prevLogProb = logProbMap.get(stringPair);
			Double logProbSum = logProbSumMap.get(stringPair.getX());

			if (prevLogProb != null) {
				Double normLogProb = prevLogProb - logProbSum;
				logProbMap.put(stringPair, normLogProb);
			}
		}
	}

	public static Set<String> makeStateSetFromTransDict(Map<StringPair, Double> transDict) {
		Set<String> stateSet = new HashSet<String>();

		for (Entry<StringPair, Double> entry : transDict.entrySet()) {
			// Add the to state for each transition.
			stateSet.add(entry.getKey().getY());
		}

		return stateSet;
	}

	/**
	 * Reads the given model parameters file and fills in the transition and emission
	 * log probabilities maps.
	 */
	public static void readModelParametersFile(FSDataInputStream in, Map<StringPair, Double> transLogProbMap,
			Map<StringPair, Double> emisLogProbMap) throws Exception {
		BufferedReader reader = new BufferedReader(new InputStreamReader(in));

		String line;
		while (null != (line = reader.readLine())) {
			EMModelParameter param = EMModelParameter.fromString(line);

			if (param != null) {
				switch (param.getParameterType()) {
				case EMModelParameter.PARAMETER_TYPE_TRANSITION:
					StringPair fromStateToState = new StringPair(param.getTransFromStateOrEmisState().toString(), param
							.getTransToStateOrEmisToken().toString());

					if (transLogProbMap.containsKey(fromStateToState)) {
						throw new Exception("Transition " + fromStateToState
								+ " was read twice in EM model parameter files.");
					}

					transLogProbMap.put(fromStateToState, param.getLogCount());
					break;
				case EMModelParameter.PARAMETER_TYPE_EMISSION:
					StringPair stateToken = new StringPair(param.getTransFromStateOrEmisState().toString(), param
							.getTransToStateOrEmisToken().toString());

					if (emisLogProbMap.containsKey(stateToken)) {
						throw new Exception("Emission " + stateToken + " was read twice in EM model parameter files.");
					}

					emisLogProbMap.put(stateToken, param.getLogCount());
					break;
				}
			}
		}

	}

}
